src/qpx_scan_algo.cpp

```diff
@@ -0,0 +1,940 @@
+    if (lba > lba_sta &&
+        (!(lba % bsize) || lba + rsize == lba_end || stat_req)) {
+      gettimeofday(&blke, NULL);
+      /*
+      btime=(blke.tv_sec - blks.tv_sec) + (blke.tv_usec -
+      blks.tv_usec)/1000000.0; spdKB = (bsize << 1) / btime; spdX  =
+      spdKB/(float)spd1X;
+      */
+      calc_cur_speed(br);
+      printf("lba: %7ld    speed: %6.2f X  %6d kB/s\r", lba, spdX, spdKB);
+#ifdef USE_FFLUSH
+      fflush(stdout);
+#endif
+      gettimeofday(&blks, NULL);
+      stat_req = 0;
+      br = 0;
+    }
+  }
+  gettimeofday(&e, NULL);
+  show_avg_speed(lba);
+  return 0;
+}
+
+#ifdef DISABLE_INTERNAL_WT
+int qscanner::run_wr_transfer() {
+  ssize_t n;
+  pid_t cpid;
+  pipe_t pipefd;
+  char linei[EMAXLINE + 1];
+  char lines[16];
+  char linef[64];
+
+  int argc;
+  char **argv;
+
+  printf("Running WRITE transfer rate test for CD/DVD/BD at speed %d...\n",
+         speed);
+
+  // creating argv...
+  argc = 0;
+  argv = (char **)malloc(sizeof(char *));
+  argv[0] = NULL;
+
+  //	execlp("cat", "cat", "/home/kgs/cdrecord.log.cd", NULL);
+  argv = add_arg(argv, &argc, "cdrecord");
+
+  sprintf(linei, "dev=%s", dev->device);
+  argv = add_arg(argv, &argc, linei);
+
+  argv = add_arg(argv, &argc, "-v");
+
+  if (WT_simul)
+    argv = add_arg(argv, &argc, "-dummy");
+
+  argv = add_arg(argv, &argc, "gracetime=2");
+  argv = add_arg(argv, &argc, "driveropts=burnfree");
+
+  sprintf(lines, "speed=%d", speed);
+  argv = add_arg(argv, &argc, lines);
+
+  if (dev->media.type & DISC_DVDpRW) {
+    sprintf(linef, "tsize=%dk", dev->media.capacity_total * 2);
+  } else {
+    sprintf(linef, "tsize=%dk", dev->media.capacity_free * 2);
+  }
+  argv = add_arg(argv, &argc, linef);
+
+  argv = add_arg(argv, &argc, "-sao");
+#ifndef _WIN32
+  argv = add_arg(argv, &argc, "/dev/zero");
+#else
+  argv = add_arg(argv, &argc, "ZERO");
+#endif
+
+  printf("qscan: starting cdrecord...\n");
+  printf("cdrecord args: \n");
+  for (int i = 0; argv[i]; i++)
+    printf("%s ", argv[i]);
+  printf("\n");
+
+  //	printf("DUMMY mode, not executing cdrecord\n");
+  //	return 0;
+
+  if ((cpid = createChildProcess(argv, &pipefd, NULL)) == -1) {
+    printf("qscan: can't create child process\n");
+    return -1;
+  }
+  printf("qscan: child created, reading from pipe...\n");
+  int wn, woffs;
+  while ((n = readline((int)pipefd[0], linei, EMAXLINE)) >= 0) {
+    //	while((n = read(pipefd[0], lineo, EMAXLINE)) > 0) {
+    //			sprintf(linei,"\nread #%d: %d bytes\n\0",idx, n);
+    //			write(sockfd, linei, strlen(linei));
+    //			idx++;
+    //				printf(lineo);
+    woffs = 0;
+    while (woffs < n) {
+      wn = write(fileno(stdout), linei + woffs, n - woffs);
+      if (wn < 0) {
+        switch (errno) {
+        case EAGAIN:
+          break;
+        default:
+          woffs = n;
+        }
+      } else {
+        woffs += wn;
+      }
+#ifdef USE_FFLUSH
+      fflush(stdout);
+#endif
+    }
+  }
+  printf("readline: %d\n", n);
```

Comment:
Codacy found an issue: [%d in format string (no. 1) requires 'int' but the argument type is 'ssize_t {aka signed long}'.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)